pkg/sources/adapter/cloudeventssource/cloudevents_test.go

```diff
@@ -0,0 +1,223 @@
+/*
+Copyright 2022 TriggerMesh Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+/*
+Contains code excerpt based on VMware's VEBA's webhook
+
+vCenter Event Broker Appliance
+Copyright (c) 2019 VMware, Inc.  All rights reserved
+
+The BSD-2 license (the "License") set forth below applies to all parts of the vCenter Event Broker Appliance project.  You may not use this file except in compliance with the License.
+
+BSD-2 License
+
+Redistribution and use in source and binary forms, with or without modification, are permitted provided that the following conditions are met:
+
+Redistributions of source code must retain the above copyright notice, this list of conditions and the following disclaimer.
+Redistributions in binary form must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.
+THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE COPYRIGHT HOLDER OR CONTRIBUTORS BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
+*/
+
+package cloudeventssource
+
+import (
+	"context"
+	"net/http"
+	"net/http/httptest"
+	"testing"
+	"time"
+
+	zapt "go.uber.org/zap/zaptest"
+
+	cloudevents "github.com/cloudevents/sdk-go/v2"
+	cloudeventst "github.com/cloudevents/sdk-go/v2/client/test"
+	"github.com/stretchr/testify/assert"
+	"github.com/stretchr/testify/require"
+	"github.com/triggermesh/triggermesh/pkg/adapter/fs"
+)
+
+const (
+	// path to secret fixtures.
+	secret1Path = "../../../../test/fixtures/secrets/secret1"
+	secret2Path = "../../../../test/fixtures/secrets/secret2"
```

Comment:
nit: In some situations it could be error prone to expect things to exist in a specific location in whatever environment is running the tests (maybe they run in a container and fixtures were not COPY'd, etc.)

I'm wondering whether the FileWatcher could work with [`fs.File`](https://pkg.go.dev/io/fs#File)? If yes, it would be trivial to [mock a file with just raw bytes](https://github.com/triggermesh/triggermesh/blob/f2928696c9d11bdc1100b0d153cfb7c0547ac4bd/hack/rbac-check/main_test.go#L73-L74) here, especially since the fixture file is literally just one word here.

Reply:
The watcher is using [fsnotify](https://github.com/fsnotify/fsnotify) which does not allow to inject a filesystem mock.
I can replace those fixtures creating `ioutil.TempFile("cloudeventsource", "secret1")` for those secrets, which I think would tackle the concern.

Do you think of a different way of solving this issue?

Reply:
Another alternative, and this one sounds better to me, is to mock the `FileWatcher`. I'll explore that for a bit.

Reply:
https://github.com/triggermesh/triggermesh/pull/730/commits/64ac555e2f9c7c4a35d31f2515e79c6bff6da83c